cli/CampaignConfigValidation/CampaignValidator.php

```diff
@@ -0,0 +1,68 @@
+<?php
+
+declare( strict_types = 1 );
+
+namespace WMDE\Fundraising\Frontend\Cli\CampaignConfigValidation;
+
+use WMDE\Fundraising\Frontend\BucketTesting\Campaign;
+use WMDE\Fundraising\Frontend\BucketTesting\CampaignCollection;
+use WMDE\Fundraising\Frontend\Cli\CampaignConfigValidation\Rule\CampaignValidationRuleInterface;
+use WMDE\Fundraising\Frontend\Cli\CampaignConfigValidation\Rule\DefaultBucketRule;
+use WMDE\Fundraising\Frontend\Cli\CampaignConfigValidation\Rule\StartAndEndTimeRule;
+use WMDE\Fundraising\Frontend\Cli\CampaignConfigValidation\Rule\UniqueBucketRule;
+
+/**
+ * Validates campaign data by looking for inconsistencies and logical errors
+ * YAML validity is checked in during loading of campaign configuration
+ *
+ * @see \WMDE\Fundraising\Frontend\BucketTesting\CampaignConfiguration
+ * @license GNU GPL v2+
+ */
+class CampaignValidator {
+
+	private $campaignCollection;
+	private $hasValidated = false;
+	private $errorLogger;
+	private $rules = [];
+
+	public function __construct( CampaignCollection $campaignCollection, ValidationErrorLogger $errorLogger ) {
+		$this->campaignCollection = $campaignCollection;
+		$this->errorLogger = $errorLogger;
+		$this->rules = [ new DefaultBucketRule(), new StartAndEndTimeRule(), new UniqueBucketRule() ];
+	}
+
+	public function isPassing(): bool {
+		return empty( $this->getErrors() );
+	}
+
+	public function getErrors(): array {
+		$this->validate();
+
+		return $this->errorLogger->getErrors();
+	}
+
+	private function validate() {
+		if ( $this->hasValidated ) {
+			return;
+		}
+
+		foreach ( $this->campaignCollection->getIterator() as $campaign ) {
+			$this->validateCampaign( $campaign );
+		}
+
+		$this->hasValidated = true;
+	}
+
+	private function validateCampaign( Campaign $campaign ): void {
+		foreach ( $this->getRules() as $rule ) {
```

Comment:
I think in this case it would be ok to use `$this->rules` directly as a performance optimization.

Reply:
The main purpose of the method is actually the PHPDoc comment as it signals that the rules should be concrete implementations of `CampaignValidationRuleInterface`. I can still remove it though / change this if you have a better idea.

Reply:
The PHPDoc could be put above the instance variable declaration in line 27. But this kind of micro-optimization is no blocker, no need to change it now.

Reply:
Found another little thing which I wanted to change so I just threw this change in as well.